src/chrome/utils/localizedError.ts

```diff
@@ -0,0 +1,119 @@
+                const textKey = messagesForFile.keys[key];
+                if (textKey !== undefined) {
+                    return textKey;
+                }
+            }
+        }
+
+        return `${file}:${key}`;
+    } else if (typeof key === 'string') {
+        return key;
+    } else {
+        return 'getNextLastKey parameter was unexpected';
+    }
+}
+
+/** Normalize the file path to the format used in the localization bundles (Relative path, no extension, and forward slashes) */
+function normalizeFilePath(file: string): string {
+    const fileRelativePath = path.relative(bundleFolderPath, file);
+    const normalizedPath = fileRelativePath.replace(/\\/g, '/').replace(/\.js$/g, '');
+    return normalizedPath;
+}
+
+const originalConfig = nls.config;
+let currentLocale: string | undefined = undefined;
+
+/** We replace nls.config with our own version so we can know which locale is set, and force all the localize functions to be re-created when needed.
+ * This works around the need to import all the secondary files only after the locale has been already set.
+ */
+(<any>nls).config = (opts?: nls.Options) => {
+    currentLocale = _.defaultTo(opts, { locale: undefined }).locale;
+    return originalConfig(opts);
+};
+
+const originalLoadMessageBundle = nls.loadMessageBundle;
+/** We replace nls.loadMessageBundle with our own custom version, which will re-created the localize function after the locale has been changed.
+ * This works around the need to import all the secondary files only after the locale has been already set.
+ */
+(<any>nls).loadMessageBundle = (file?: string) => {
+    let localizeForFile = originalLoadMessageBundle(file);
+    let localizeForFileLocale = currentLocale;
+    return (key: string | LocalizeInfo, message: string, ...args: (string | number | boolean | undefined | null)[]) =>
+        {
+            // Has the locale changed since we created the localize function?
+            if (localizeForFileLocale !== currentLocale) {
+                localizeForFileLocale = currentLocale;
+                // If so, re-create it so it'll use the new local (the localize caches the local, so we do need to re-create it after changing the locale)
+                localizeForFile = originalLoadMessageBundle(file);
+            }
+
+            // Call our custom localize which will store the localization key for the LocaliedErrors
+            return customLocalize(file, localizeForFile, <string>key, message, ...args);
+        };
+};
+
+// Consume the key of the latest call to the localize function
+function consumeLastLocalizationKey(): string {
+    const lastKeyToReturn = lastKeyOfLocalize;
+    lastKeyOfLocalize = defaultLastKeyOfLocalize;
+    return lastKeyToReturn;
+}
+
+/** Error class which automatically extracts the localization key of the latest call to localize (So we can easily deduplicate errors on telemetry) */
+export class LocalizedError extends Error {
```

Comment:
I'm not sure the error code should be tied to the localized string id at all. These are two different things and it seems like they should operate independantly. Maybe we will want to change the error code without losing the localized string, or have multiple errors share the same string, or multiple strings associated with one error, or have an error code without a string attached, etc etc. Plus this behavior is hidden and non-obvious.

Reply:
That's a good point. Changing to a different approach so we don't tied the error-id and the localization-id.